### ISSUE TITLE
CI: have some mechanism to check that all-inclusive header does include every non-internal header for a given library

### ISSUE THREAD

**Issue by antkmsft** [Azure/azure-sdk-for-cpp#4234]
It is only a matter of time until we add a new header, but forget to update the all-inclusive header (i.e. `/azure/core.hpp`).
There have been cases in the past. Maybe we already have one again.

Having this mechanism is good for us, so we won't make this mistake. It is good for our customers.
It is also good for the assumptions that we'll be able to make, i.e. ApiView can get away with processing only that one single header, without implementing wildcard support.

This probably can be implemented as a powershell script that runs in CI.

Once we start implementing, it may turn out that we don't want some header in the all-inclusive header for some valid reasons. Then, there would be an exclusion mechanism - it can be a file with exceptions, or we can agree on a format that all-inclusive headers would have a comment for each header that it does not include, i.e. `// not including: azure/core/some_header.hpp`, this way there will be a comment nearby, and everything will be kept in one file.
What I like in this approach is that not including headers will be explicit - it won't be possible to not include header simply by mistake.

cc @ahsonkhan, cc @LarryOsterman - I saw you were talking about this problem in the context of ApiView and I thought this might be a solution.

**Comment by LarryOsterman** [comment]
IMHO this is a great idea, but it's important to remember that the all-inclusive header should only contain the public types and APIs which are universally available.

That means that types in the _internal namespace (which are public but are only intended for consumption by service packages) and types which are conditionally available (WinHTTP transport) shouldn't be in the all-inclusive header.

That makes the tool slightly trickier to implement, but it is still possible.

**Comment by antkmsft** [comment]
Ah, yes... If we want `_internal` namespace in ApiView, then this tool is probably not the best approach to do that for ApiView.
I still like the idea, for the initial reasons of not forgetting to update the all-inclusive header, it just probably won't be an elegant solution for the ApiView's purposes as its side effect.

**Comment by LarryOsterman** [comment]
> Ah, yes... If we want `_internal` namespace in ApiView, then this tool is probably not the best approach to do that for ApiView. I still like the idea, for the initial reasons of not forgetting to update the all-inclusive header, it just probably won't bring that elegant solution for the ApiView's purposes as its side effect.

There are several components (azure-core-cpp and azure-storage-common-cpp specifically) which publish types in the _internal namespace for consumption to other packages.

Because these types in the _internal namespace are consumed by components outside the package, I think they should be included in the ApiView because the _internal types are a part of the public contract for the package.

As such, I think that they should be a part of the API review (more importantly, except for the two packages mentioned, nobody should be publishing any types in the _internal namespace, including _internal types brings visibility to those errors).

This is different from types in the _detail namespace, which by definition are NOT a part of the public API surface for the SDK.

And to be clear: I can easily see an argument that says that we should exclude all _internal types from the API view.

**Comment by antkmsft** [comment]
Yes, such argument can be made, given that other SDKs don't seem to include internal types, but it does not necessarily mean it is the best approach.
